### PR TITLE
Beejones/key release policy operators

### PR DESCRIPTION
**Pull request by beejones**
Before this change, the key release policy was only using an array of claims which needed to be found in the attestation.
The policy is extended with operators. We added two operators:
gt: Greater than indicating that the attestation claim must be greater than what is indicated in the policy. 
gte: Greater than or equal.

The key release policy logic changed completely as also the constitution.
New unit tests check the logic of key release policy.

### CODE REVIEW

**Comment by beejones** [comment]
Tests:
https://github.com/microsoft/privacy-sandbox-dev/actions/runs/11036104516

File: test/e2e-test/src/index.ts [hunk 1]
```diff
@@ -515,16 +515,16 @@ class Demo {
...
-      `keyResponse["x-ms-ver"][0] === '2'`,
-      keyResponse["x-ms-ver"][0] === "2",
+      `keyResponse.claims["x-ms-ver"][0] === '2'`,
+      keyResponse.claims["x-ms-ver"][0] === "2",
     );
     Demo.assert(
-      `keyResponse["x-ms-sevsnpvm-is-debuggable"][0] === false`,
-      keyResponse["x-ms-sevsnpvm-is-debuggable"][0] === false,
+      `keyResponse.claims["x-ms-sevsnpvm-is-debuggable"][0] === false`,
+      keyResponse.claims["x-ms-sevsnpvm-is-debuggable"][0] === false,
```
Comment (takuro-sato) [inline test/e2e-test/src/index.ts:527]: How about having some basic test cases for gt and gte?

File: test/unit-test/policies/IKeyReleasePolicySnpProps.test.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+    // Assert
+    expect(validationResult.success).toBe(true);
+  });
+  test("Should validate successfully Key Release Policy properties with operator gte", () => {
+    // Arrange
+    Logger.setLogLevel(LogLevel.DEBUG);
+    const policy: any = {
+      type: "",
+      claims: { "x-ms-attestation-type": ["sevsnpvm"] },
```
Comment (takuro-sato) [inline test/unit-test/policies/IKeyReleasePolicySnpProps.test.ts:51]: For better coverage, can we have a case where a policy filed have multiple values something like `"x-ms-attestation-type": ["sevsnpvm", "some-other-value"]`?

File: src/policies/IKeyReleasePolicySnpProps.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+export interface IKeyReleasePolicySnpProps {
+  "x-ms-attestation-type"?: string[] | string | string;
+  "x-ms-compliance-status"?: string[] | string;
```
[the remark this answers is not in the document]
Comment (takuro-sato) [reply src/policies/IKeyReleasePolicySnpProps.ts:6]: Also from governance/constitution/kms_actions.js, looks like we have assumption that the values are array.
```
let item = claims[key];
          // Make sure item is always an array
          if (!Array.isArray(item)) {
            item = [item];
          }

```

File: governance/policies/key-release-policy-add.json
Comment (takuro-sato) [inline governance/policies/key-release-policy-add.json:1]: Can we have a test for updating a policy with a running KMS instance?
Something like:
- Create a first policy with:
```
"claims": {
          "x-ms-ver": ["2", "3"]
        },
```
- Remove `claim`, then add
```
"gte": {
          "x-ms-ver": "2"
        },
```
- get policy from the endpoint and check if the policy value is what we expect

File: scripts/setup_mCCF.sh [hunk 4]
```diff
@@ -5,7 +5,9 @@
 
 set -euo pipefail
 export AUTHORIZATION="Bearer $ACCESS"
-export CCF_NAME="acceu-..."
+export AUTHORIZATION=$(./scripts/authorization_header.sh)
+#export CCF_NAME="acceu-bingads-502-1"
+export CCF_NAME="trust-coordinator-na-504"
```
[the remark this answers is not in the document]
Comment (beejones) [reply scripts/setup_mCCF.sh:10]: accidental checkin. will revert.

[the remark this answers is not in the document]
Comment (beejones) [reply scripts/setup_mCCF.sh:10]: done

File: src/policies/IKeyReleasePolicySnpProps.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+export interface IKeyReleasePolicySnpProps {
+  "x-ms-attestation-type"?: string[] | string | string;
```
[the remark this answers is not in the document]
Comment (beejones) [reply src/policies/IKeyReleasePolicySnpProps.ts:5]: done

File: governance/policies/key-release-policy-remove.json [hunk 6]
```diff
@@ -7,6 +7,12 @@
         "type": "remove",
         "claims": {
           "x-ms-ver": ["2", "3"]
+        },
+        "gte": {
+          "x-ms-ver": "2"
+        },
+        "gt": {
+          "x-ms-ver": "1"
```
[the remark this answers is not in the document]
Comment (beejones) [reply governance/policies/key-release-policy-remove.json:15]: In this way we always test all operators. These operators are in the add policy.

File: test/e2e-test/src/index.ts (hunk 1 above)
Comment (beejones) [reply test/e2e-test/src/index.ts:527]: we have many in the unit tests

File: governance/policies/key-release-policy-add.json
Comment (beejones) [reply governance/policies/key-release-policy-add.json:1]: I created an issue for this one: https://github.com/microsoft/azure-privacy-sandbox-kms/issues/139

File: governance/constitution/kms_actions.js [hunk 7]
```diff
@@ -117,116 +117,254 @@ actions.set(
...
+        // get all the claims for type from the KV
+        let keyBuf = ccf.strToBuf(type);
+        if (ccf.kv[keyReleaseMapName].has(keyBuf)) {
+          // type is already available
+          const itemsBuf = ccf.kv[keyReleaseMapName].get(keyBuf);
+          items = ccf.bufToStr(itemsBuf);
+          console.log(
+            `KRP add ${type}=>key: ${type} already exist: ${items} in the key release policy`,
+          );
+          items = JSON.parse(items);
```
[the remark this answers is not in the document]
Comment (beejones) [reply governance/constitution/kms_actions.js:134]: done

[the remark this answers is not in the document]
Comment (beejones) [reply governance/constitution/kms_actions.js:134]: done

File: src/attestation/AttestationValidation.ts [hunk 8]
```diff
@@ -125,7 +126,7 @@ export const validateAttestation = (
       }`,
     );
 
-    if (Object.keys(keyReleasePolicy).length === 0) {
+    if (Object.keys(keyReleasePolicy.claims).length === 0) {
```
[the remark this answers is not in the document]
Comment (beejones) [reply src/attestation/AttestationValidation.ts:129]: done

File: src/endpoints/keyReleasePolicyEndpoint.ts [hunk 9]
```diff
@@ -17,14 +18,15 @@ enableEndpoint();
...
   const serviceRequest = new ServiceRequest<void>(name, request);
 
   // check if caller has a valid identity
   const [_, isValidIdentity] = serviceRequest.isAuthenticated();
   if (isValidIdentity.failure) return isValidIdentity;
 
-  const result = getKeyReleasePolicy(keyReleasePolicyMap);
-  return ServiceResult.Succeeded<IKeyReleasePolicyProps>(result);
+  const result =
+    KeyReleasePolicy.getKeyReleasePolicyFromMap(keyReleasePolicyMap);
```
[the remark this answers is not in the document]
Comment (beejones) [reply src/endpoints/keyReleasePolicyEndpoint.ts:30]: done

File: test/unit-test/policies/IKeyReleasePolicySnpProps.test.ts [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    policy["x-ms-attestation-type"] = ["sevsnpvm", "none"];
+
+    // Assert
+    expect(policy["x-ms-attestation-type"]).toContain("sevsnpvm");
+    expect(policy["x-ms-attestation-type"]).toContain("none");
+  });
+
+  test("Should validate successfully", () => {
+    // Arrange
+    Logger.setLogLevel(LogLevel.DEBUG);
```
[the remark this answers is not in the document]
Comment (beejones) [reply test/unit-test/policies/IKeyReleasePolicySnpProps.test.ts:28]: done

File: test/unit-test/policies/IKeyReleasePolicySnpProps.test.ts (hunk 2 above)
Comment (beejones) [reply test/unit-test/policies/IKeyReleasePolicySnpProps.test.ts:51]: that is the case:
policy["x-ms-attestation-type"] = ["sevsnpvm", "none"];

    // Assert
    expect(policy["x-ms-attestation-type"]).toContain("sevsnpvm");
    expect(policy["x-ms-attestation-type"]).toContain("none");

    Did I miss the point?

File: src/policies/IKeyReleasePolicy.ts [hunk 11]
```diff
@@ -1,9 +1,11 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
-import { IKeyReleasePolicyProps } from "..";
+import { IKeyReleasePolicySnpProps } from "..";
 
 export interface IKeyReleasePolicy {
   type: string;
```
[the remark this answers is not in the document]
Comment (beejones) [reply src/policies/IKeyReleasePolicy.ts:7]: done

File: src/policies/IKeyReleasePolicySnpProps.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (beejones) [reply src/policies/IKeyReleasePolicySnpProps.ts:6]: IKeyReleasePolicySnpProps is used for claims and operators such as gte. Claims can be an array but operators not.

File: src/policies/KeyReleasePolicy.ts [hunk 12]
```diff
@@ -1,11 +1,250 @@
...
+      const kvValueBuf = keyReleasePolicyMap.get(kvKeyBuf);
+      if (!kvValueBuf) {
+        if (!kv.optinal) {
+          throw new Error(
+            `Key release policy ${kvKey} not found in the key release policy map`,
+          );
+        }
+      } else {
+        let kvValue = ccf.bufToStr(kvValueBuf!);
+        keyReleasePolicy[kvKey] = JSON.parse(
```
[the remark this answers is not in the document]
Comment (beejones) [reply src/policies/KeyReleasePolicy.ts:241]: I catch the parse.

File: test/e2e-test/src/index.ts (hunk 1 above)
Comment (beejones) [reply test/e2e-test/src/index.ts:527]: done

**Comment by beejones** [comment]
new test:
https://github.com/microsoft/privacy-sandbox-dev/actions/runs/11067938914

File: test/unit-test/policies/IKeyReleasePolicySnpProps.test.ts (hunk 2 above)
Comment (takuro-sato) [reply test/unit-test/policies/IKeyReleasePolicySnpProps.test.ts:57]: Can we call `KeyReleasePolicy.validateKeyReleasePolicy` I don't see it in the test?

File: src/policies/KeyReleasePolicy.ts [hunk 13]
```diff
@@ -1,11 +1,250 @@
...
+      );
+    }
+    if (attestationClaims === null || attestationClaims === undefined) {
+      return ServiceResult.Failed<string>(
+        { errorMessage: "Missing attestation claims" },
+        500,
+      );
+    }
+
+    const claims = keyReleasePolicyClaims;
```
[the remark this answers is not in the document]
Comment (takuro-sato) [reply src/policies/KeyReleasePolicy.ts:38]: What do you think about it?

File: src/policies/KeyReleasePolicy.ts [hunk 14]
```diff
@@ -1,11 +1,250 @@
...
+          {
+            errorMessage: `Policy value for claim ${key} is not a number or a string representing a number for operator type ${type}`,
+          },
+          400,
+        );
+      }
+
+      if (typeof policyValue !== "number") {
+        policyValue = parseFloat(policyValue);
+      }
```
[the remark this answers is not in the document]
Comment (takuro-sato) [reply src/policies/KeyReleasePolicy.ts:134]: What do you think about it?

[the remark this answers is not in the document]
Comment (beejones) [reply src/policies/KeyReleasePolicy.ts:134]: gt/gte should only be used on numbers

[the remark this answers is not in the document]
Comment (takuro-sato) [reply src/policies/KeyReleasePolicy.ts:132]: Yeah, I know that's the current implementation.
Do you know which claims are actually required to handle gt, gte? And what's the value type for these claims? We are really sure that they are all simple numbers?

My point is that if there is a claim `version-for-something` and the value looks like '1.0.1' and it's required to verified with gt/gte, our implementation wound not meet the requirement.

[the remark this answers is not in the document]
Comment (takuro-sato) [reply src/policies/KeyReleasePolicy.ts:132]: We don't know the requirement yet and will get feedback

**Review by takuro-sato — APPROVED** [review]

**Pull request merged**